File: 73_Reverse/javascript/reverse.js
// REVERSE
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;
    
    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");
                       
                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

var a = [];
var n;

// Subroutine to print the rules
function print_rules()
{
    print("\n");
    print("THIS IS THE GAME OF 'REVERSE'.  TO WIN, ALL YOU HAVE\n");
    print("TO DO IS ARRANGE A LIST OF NUMBERS (1 THROUGH " + n + ")\n");
    print("IN NUMERICAL ORDER FROM LEFT TO RIGHT.  TO MOVE, YOU\n");
    print("TELL ME HOW MANY NUMBERS (COUNTING FROM THE LEFT) TO\n");
    print("REVERSE.  FOR EXAMPLE, IF THE CURRENT LIST IS:\n");
    print("\n");
    print("2 3 4 5 1 6 7 8 9\n");
    print("\n");
    print("AND YOU REVERSE 4, THE RESULT WILL BE:\n");
    print("\n");
    print("5 4 3 2 1 6 7 8 9\n");
    print("\n");
    print("NOW IF YOU REVERSE 5, YOU WIN!\n");
    print("\n");
    print("1 2 3 4 5 6 7 8 9\n");
    print("\n");
    print("NO DOUBT YOU WILL LIKE THIS GAME, BUT\n");
    print("IF YOU WANT TO QUIT, REVERSE 0 (ZERO).\n");
    print("\n");
}

// Subroutine to print list
function print_list()
{
    print("\n");
    for (k = 1; k <= n; k++)
        print(" " + a[k] + " ");
    print("\n");
    print("\n");
}

// Main program
async function main()
{
    print(tab(32) + "REVERSE\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    print("REVERSE -- A GAME OF SKILL\n");
    print("\n");
    for (i = 0; i <= 20; i++)
        a[i] = 0;
    // *** N=NUMBER OF NUMBER
    n = 9;
    print("DO YOU WANT THE RULES");
    str = await input();
    if (str != "NO")
        print_rules();
    while (1) {
        // *** Make a random list a(1) to a(n)
        a[1] = Math.floor((n - 1) * Math.random() + 2);
        for (k = 2; k <= n; k++) {
            do {
                a[k] = Math.floor(n * Math.random() + 1);
                for (j = 1; j <= k - 1; j++) {
                    if (a[k] == a[j])
                        break;
                }
            } while (j <= k - 1) ;
        }
        // *** Print original list and start game
        print("\n");
        print("HERE WE GO ... THE LIST IS:\n");
        t = 0;
        print_list();
        while (1) {
            while (1) {
                print("HOW MANY SHALL I REVERSE");
                r = parseInt(await input());
                if (r == 0)
                    break;
                if (r <= n)
                    break;
                print("OOPS! TOO MANY! I CAN REVERSE AT MOST " + n + "\n");
            }
            if (r == 0)
                break;
            t++;
            // *** Reverse r numbers and print new list
            for (k = 1; k <= Math.floor(r / 2); k++) {
                z = a[k];
                a[k] = a[r - k + 1];
                a[r - k + 1] = z;
            }
            print_list();
            // *** Check for a win
            for (k = 1; k <= n; k++) {
                if (a[k] != k)
                    break;
            }
            if (k > n) {
                print("YOU WON IT IN " + t + " MOVES!!!\n");
                print("\n");
                break;
            }
        }
        print("\n");
        print("TRY AGAIN (YES OR NO)");
        str = await input();
        if (str != "YES")
            break;
    }
    print("\n");
    print("O.K. HOPE YOU HAD FUN!!\n");
}

main();

File: 73_Reverse/ruby/reverse.rb
ARRAYSIZE = 9
$digitArray = Array.new(ARRAYSIZE)
$winningArray = Array.new(ARRAYSIZE)

# Method to print the rules
def displayTheRules
    puts "This is the game of 'Reverse'.  to win, all you have"
    puts "to do is arrange a list of numbers (1 through " + ARRAYSIZE.to_s + ")"
    puts "in numerical order from left to right.  to move, you"
    puts "tell me how many numbers (counting from the left) to"
    puts "reverse.  For example, if the current list is:"
    puts "2 3 4 5 1 6 7 8 9"
    puts "and you reverse 4, the result will be:"
    puts "5 4 3 2 1 6 7 8 9"
    puts "Now if you reverse 5, you win!"
    puts "1 2 3 4 5 6 7 8 9"
    puts "No doubt you will like this game, but"
    puts "if you want to quit, reverse 0 (zero)."
end

# Method to print the list
def printList
    puts "\n" + $digitArray.join(" ") + "\n\n"
end

# Zero-based arrays contain digits 1-9
# Make a random array and an ordered winning answer array A[0] to A[N]
def makeRandomList
    for kIndex in 0..ARRAYSIZE-1 do
        $digitArray[kIndex] = kIndex+1
        $winningArray[kIndex] = kIndex+1
    end
    # now randomize the digit array order
    $digitArray.shuffle!
end

def checkForWin? (triesSoFar)
    # Check for a win (all array elements in order)
    if $digitArray == $winningArray then
        puts "You won it in " + triesSoFar.to_s + " moves!!!\n\n"
        puts "try again (yes or no)?"
        tryAgain = gets.strip.upcase
        if tryAgain == "YES" then
            return true
        end
        puts "\nO.K. Hope you had fun!!"
        exit
    end
    return false
end

def reverseIt (howManyToReverse, triesSoFar)
    # REVERSE R NUMBERS AND PRINT NEW LIST

    # extract and reverse the first howManyToReverse elements of the array
    subArray = $digitArray.take(howManyToReverse)
    subArray.reverse!
    
    # get the remaining elements of the original array
    endArray = $digitArray.slice(howManyToReverse, ARRAYSIZE)
    # append those elements to the reversed elements
    $digitArray = subArray.concat(endArray)

    # if we got all in order, randomize again
    isWinner = checkForWin?(triesSoFar)
    if isWinner == true then
        makeRandomList
    end
    printList # always print the newly ordered list
    return isWinner
end

def askHowManyToReverse
    puts "How many shall I reverse?";
    rNumber = gets.to_i
    if rNumber > 0 then
        if rNumber > ARRAYSIZE then
            puts "Oops! Too many! I can reverse at most " + ARRAYSIZE.to_s
        end
    else
        rNumber = 0 # zero or negative values end the game
    end
    return rNumber
end

puts "REVERSE"
puts "Creative Computing  Morristown, New Jersey\n\n\n"
puts "REVERSE -- A game of skill\n\n"

puts "Do you want the rules?"
wantRules = gets.strip.upcase
if wantRules == "YES" then
    displayTheRules
end

makeRandomList
howManyTries = 0
puts "\nHere we go ... the list is:"
printList # display the initial list
# start the game loop
r = askHowManyToReverse
while r != 0 do # zero will end the game
    if r <= ARRAYSIZE then
        howManyTries = howManyTries+1
        if reverseIt(r, howManyTries) then
            howManyTries = 0
        end
    end
    r = askHowManyToReverse
end

